### ISSUE TITLE
Feature Request: Add support for extern aliases

### ISSUE THREAD

**Issue by siegfriedpammer** [icsharpcode/ILSpy#2442]
### Discussed in https://github.com/icsharpcode/ILSpy/discussions/2441

<div type='discussions-op-text'>

<sup>Originally posted by **aBaechtold** June 24, 2021</sup>
Hi,

I was recently reviewing an assembly that relies on assembly aliases  (see also [extern alias (C# Reference)](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/extern-alias)) to access types in assemblies having the same fully qualified name but of different version. I noticed that ILSpy displays the types of the different versions to look exactly the same i.e. couldn't tell that they were different while they were. Clicking each of them with the mouse pointer navigated to the correct version, so that is good.

I wondered whether this could be an improvement and before submitting an issue wanted to ask if others think similarly. I did not find an issue or discussion that seems to cover this topic already, hence the new discussion. It's nothing critical as using assembly aliases is not a common thing I guess. It's also not showing something incorrectly, it's just ambiguity in the UI.

### Details:
ILSpy version 7.1.0.6543
.NET version 4.8.4300.0
 
**ILSpy disassembly:**
```
private void _GenerateListOfConfigModels()
	{
		List<VConfigInfo> list = new List<VConfigInfo>
		{
			new VConfigInfo
			{
				VersionString = new VConfig().ExpectedConfigVersion,
				ConfigType = typeof(VConfig)   //<=Here...
			},
			new VConfigInfo
			{
				VersionString = new VConfig().ExpectedConfigVersion,
				ConfigType = typeof(VConfig)  //<=and here
			}
		};
		...
	}
```

**Actual code:**
```
...
extern alias v1;
extern alias v2;
...
using VConfig_v1               = v1::VConfig;
using VConfig_v2               = v2::VConfig;
...
private void _GenerateList() 
   {
      List<VConfigInfo> list = new List<VConfigInfo>
          {
               new VConfigInfo {VersionString = new VConfig_v1().ExpectedConfigVersion, ConfigType = typeof(VConfig_v1)},
               new VConfigInfo {VersionString = new VConfig_v2().ExpectedConfigVersion, ConfigType = typeof(VConfig_v2)}
                ...
          };
     ...
    }
```
</div>

**Comment by siegfriedpammer** [comment]
Thank you for the suggestion! Currently there is no support for extern aliases in ILSpy's decompiler engine, although the type system and syntax tree have some kind of support for this. Would you happen to know whether the fact that some assemblies have an extern alias is somehow manifested in the compiled assembly/metadata?

If not, ILSpy would have to detect whether there are types in different assemblies that have the exact same full type name. Adding full support for extern aliases does seem like a fair amount of work. Not sure, when we will be able to add this.

Of course, after an implementation has been specified, this could be contributed by anyone.

**Comment by aBaechtold** [comment]
Hi Siegfried,
Sorry, I missed your question. 
Unfortunately I don't know if that aspect is reflected in the assemblies meta data somehow.
I'd need to do some research. I'll try to follow up on that.

**Comment by aBaechtold** [comment]
I did a bit of testing and it looks like the exact alias is not retained in the assembly (meta data).
There will be just multiple assembly references (AssemblyRef) that are the same despite the version part and the code will reference the corresponding parent. In such situations, what one could do is to just generate some clever alias (whatever clever means in that context).
An alias in IL is present by means of 'as' in the AssemblyRef (see ECMA-335).
 
ildasm does something like that. Here's an example:
![image](https://user-images.githubusercontent.com/20663491/141195864-8881f6b5-90a8-49ce-a1aa-e18a5de8b112.png)
![image](https://user-images.githubusercontent.com/20663491/141195909-4440743e-bec8-45fc-89dc-cd30e0a29dd6.png)

Is there some way I can share my test console application (for which the above screenshots were taken) ?

**Comment by siegfriedpammer** [comment]
> Is there some way I can share my test console application (for which the above screenshots were taken) ?
 
You can just attach the file to an issue comment or please send it to siegfried.pammer@gmail.com. Thanks!

**Comment by aBaechtold** [comment]
After some trial an error to provide you the files (both as comment attachments or via email - Gmail declined all attachments) figured out that I should more carefully read the error message when uploading data to the comment. The issue was that I used 7Z or TAR archives which are not supported, but ZIP is. Thus here we go with the ZIP file :-)
[ExampleConsoleApp.zip](https://github.com/icsharpcode/ILSpy/files/7521801/ExampleConsoleApp.zip)

**Comment by aBaechtold** [comment]
I could try to provide a suggestion (code changes) from my end, but might take a while...

**Comment by siegfriedpammer** [comment]
A quick list of TODOs (from the top of my head):

* [] `ReflectionDisassmbler`: check if there are name collisions in the list of references, and generate aliases of the form `assembly.name_N` where N is a counter >= 2 as needed, also make sure to use these aliases when printing member references.
* [] type system: `MetadataModule` might need extensions to support extern aliases
* [] the `IntroduceUsingDeclarations` transform will have to emit `extern alias` directives and apply them in the decompiled code

**Comment by siegfriedpammer** [comment]
> I could try to provide a suggestion (code changes) from my end, but might take a while...

If you want to implement this feature, let me know... I can help you, if you have any questions. Thanks!

**Comment by aBaechtold** [comment]
If you don't mind me being slow, then sure, I'd like to help if I can. As far as I can judge this is rather an exotic edge case not commonly found and the existing behavior is not broken, it's just not optimal, so a quick implementation might not be required.

So far I cloned the repository and was able to compile the solution. I will need to setup local code formatting to be as you outlined (tabs). 

I was a bit overwhelmed by the existing code and looked in the Wiki if there are any guidelines or design concept descriptions, but have not exactly found something like that. 
I take it that I could use your TODO list above as a pointer into the code.

**Comment by siegfriedpammer** [comment]
> I take it that I could use your TODO list above as a pointer into the code.

Yes, sorry for the non-existing documentation. If you have any questions or get stuck somewhere, please feel free to ask.
